debian/parrot_debugger.1
.\" Automatically generated by Pod::Man 2.1801 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PARROT_DEBUGGER.C 1"
.TH PARROT_DEBUGGER.C 1 "2009-03-08" "Debian Project" "Debian GNU/Linux manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
parrot_debugger \- The Parrot debugger
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& parrot_debugger programfile
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.SS "Commands"
.IX Subsection "Commands"
.ie n .IP """disassemble""" 4
.el .IP "\f(CWdisassemble\fR" 4
.IX Item "disassemble"
Disassemble the bytecode.
.Sp
Use this if you have a \s-1PBC\s0 file but not the \s-1PASM\s0.
.ie n .IP """load""" 4
.el .IP "\f(CWload\fR" 4
.IX Item "load"
Load a source code file.
.ie n .IP """list"" or ""l""" 4
.el .IP "\f(CWlist\fR or \f(CWl\fR" 4
.IX Item "list or l"
List the source code file.
.ie n .IP """run"" or ""r""" 4
.el .IP "\f(CWrun\fR or \f(CWr\fR" 4
.IX Item "run or r"
Run the program.
.ie n .IP """break"" or ""b""" 4
.el .IP "\f(CWbreak\fR or \f(CWb\fR" 4
.IX Item "break or b"
Add a breakpoint.
.ie n .IP """watch"" or ""w""" 4
.el .IP "\f(CWwatch\fR or \f(CWw\fR" 4
.IX Item "watch or w"
Add a watchpoint.
.ie n .IP """delete"" or ""d""" 4
.el .IP "\f(CWdelete\fR or \f(CWd\fR" 4
.IX Item "delete or d"
Delete a breakpoint.
.ie n .IP """disable""" 4
.el .IP "\f(CWdisable\fR" 4
.IX Item "disable"
Disable a breakpoint.
.ie n .IP """enable""" 4
.el .IP "\f(CWenable\fR" 4
.IX Item "enable"
Reenable a disabled breakpoint.
.ie n .IP """continue"" or ""c""" 4
.el .IP "\f(CWcontinue\fR or \f(CWc\fR" 4
.IX Item "continue or c"
Continue the program execution.
.ie n .IP """next"" or ""n""" 4
.el .IP "\f(CWnext\fR or \f(CWn\fR" 4
.IX Item "next or n"
Run the next instruction
.ie n .IP """eval"" or ""e""" 4
.el .IP "\f(CWeval\fR or \f(CWe\fR" 4
.IX Item "eval or e"
Run an instruction.
.ie n .IP """trace"" or ""t""" 4
.el .IP "\f(CWtrace\fR or \f(CWt\fR" 4
.IX Item "trace or t"
Trace the next instruction.
.ie n .IP """print"" or ""p""" 4
.el .IP "\f(CWprint\fR or \f(CWp\fR" 4
.IX Item "print or p"
Print the interpreter registers.
.ie n .IP """stack"" or ""s""" 4
.el .IP "\f(CWstack\fR or \f(CWs\fR" 4
.IX Item "stack or s"
Examine the stack.
.ie n .IP """info""" 4
.el .IP "\f(CWinfo\fR" 4
.IX Item "info"
Print interpreter information.
.ie n .IP """quit"" or ""q""" 4
.el .IP "\f(CWquit\fR or \f(CWq\fR" 4
.IX Item "quit or q"
Exit the debugger.
.ie n .IP """help"" or ""h""" 4
.el .IP "\f(CWhelp\fR or \f(CWh\fR" 4
.IX Item "help or h"
Print the help.
.SS "Debug Ops"
.IX Subsection "Debug Ops"
You can also debug Parrot code by using the \f(CW\*(C`debug_init\*(C'\fR, \f(CW\*(C`debug_load\*(C'\fR
and \f(CW\*(C`debug_break\*(C'\fR ops in \fIops/debug.ops\fR.
.ie n .IP """int main(int argc, char *argv[])""" 4
.el .IP "\f(CWint main(int argc, char *argv[])\fR" 4
.IX Item "int main(int argc, char *argv[])"
Reads the \s-1PASM\s0 or \s-1PBC\s0 file from argv[1], loads it, and then calls
\&\fIParrot_debug()\fR.
.ie n .IP """static void PDB_add_exception_handler(Parrot_Interp)""" 4
.el .IP "\f(CWstatic void PDB_add_exception_handler(Parrot_Interp)\fR" 4
.IX Item "static void PDB_add_exception_handler(Parrot_Interp)"
Adds a default exception handler to \s-1PDB\s0.
.ie n .IP """static void PDB_printwelcome(void)""" 4
.el .IP "\f(CWstatic void PDB_printwelcome(void)\fR" 4
.IX Item "static void PDB_printwelcome(void)"
Prints out the welcome string.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIsrc/debug.c\fR, \fIinclude/parrot/debug.h\fR.
.SH "HISTORY"
.IX Header "HISTORY"
.IP "\(bu" 4
Initial version by Daniel Grunblatt on 2002.5.19.
.IP "\(bu" 4
Start of rewrite \- leo 2005.02.16
.Sp
The debugger now uses it's own interpreter. User code is run in
Interp* debugee. We have:
.Sp
.Vb 4
\&  debug_interp\->pdb\->debugee\->debugger
\&    ^                            |
\&    |                            v
\&    +\-\-\-\-\-\-\-\-\-\-\-\-\- := \-\-\-\-\-\-\-\-\-\-\-+
.Ve
.Sp
Debug commands are mostly run inside the \f(CW\*(C`debugger\*(C'\fR. User code
runs of course in the \f(CW\*(C`debugee\*(C'\fR.
.SH "TODO"
.IX Header "TODO"
.IP "\(bu" 4
Check the user input for bad commands, it's quite easy to make
it bang now, try listing the source before loading or disassembling it.
.IP "\(bu" 4
Print the interpreter info.
.IP "\(bu" 4
Make the user interface better (add comands
history/completion).
.IP "\(bu" 4
Some other things I don't remember now because it's late.


NEWS
# $Id: NEWS 19133 2007-06-19 17:07:42Z allison $

New in 0.4.13
- Languages:
 + Updated Lisp, Lua, PHP ("Plumhead"), Python ("Pynie"), ABC,
   WMLScript, and Tcl ("ParTcl").
 + Perl 6 passes all of the sanity tests.
 + PGE supports latest Perl 6 grammar syntax. Perl 6, Python
   ("Pynie"), and ABC parsers updated to match.
 + Updated PHP ("Plumhead") to Antlr 3.0.
 + Lua added the beginnings of a PGE/TGE based compiler (not yet
   replacing the Perl/Yapp compiler).
 + Lisp updated for current features, added a test suite.
- Core Implementation:
 + Filled in features and backward compatibility for PDD 15 objects.
   New object metamodel passes 85% of old test suite.
 + GCC API symbols are visible externally only when explicitly exported.
 + Added generated GCC compiler attributes to increase warnings, and
   cleaned up resulting warnings.
 + Code cleanup efforts and fixed memory leaks by the cage cleaners,
   resulting in notable speed increases.
- Misc:
 + Updated Parrot distribution to Artistic License 2.0, from dual
   Artistic 1/GPL license.
 + SDL examples brought up-to-date with current features.

New in 0.4.12
- Build:
 + even more refactorings and improvements in configuration system
 + improved tests and documentation for configuration system
- Languages:
 + Updated abc, PHP ("Plumhead"), Lua, Tcl
 + Reclaimed Lisp
- Design:
 + new PMC PDD draft partially completed
- Implementation:
 + continued implementation of PDD 15 (Objects)
 + minor garbage collector and memory management improvements
 + several memory leaks resolved
 + extended support for compilers other than GCC
 + improved C++ compiler compatibility
 + enhanced support for Solaris, non-MSVC Win32, and other platforms

New in 0.4.11

- Compilers:
 + IMCC: added documentation for C-based Parrot Calling Conventions,
   refactorings and bug fixes
 + PGE: new perl6regex front end reflecting recent S05 syntax changes
 + PIRC: new prototype PIR parser
- Languages:
 + Updated Lua, PHP ("Plumhead"), BASIC, pynie
 + Lua implements environment
- Design:
 + PDD15 "Objects" - details added, and draft approved
- Documentation:
 + Added guidelines for PMC documentation
- Implementation:
 + PDD15 implementation is largely complete, including role-based composition,
   introspection, and C3 method resolution order
 + new Exporter PMC for importing globals between namespaces
 + new string utilities for radix conversion
 + PCCINVOKE and Parrot_PCCINVOKE allow calling using the full Parrot Calling
   Conventions from PMCs and C code respectively
- Build:
 + Refactorings and improvements in test coverage for 'Configure.pl'
- Misc:
 + many bugfixes, enhancements, and code cleanup
 + added example subversion config file
 + extended support for gcc, icc, and other compilers
 + extended support for Solaris and other platforms


New in 0.4.10

- Compilers:
 + IMCC: Parrot calling conventions now available between two C PMCs (PMINVOKE)
 + PGE: Match object improvements
 + smop: added Attribute and Class PMCs
 + HLLCompiler: improvements for interactive mode
- PAST:
 + extended binding to a list of variables
- Languages:
 + Updated Lua, PHP ("Plumhead"), Tcl ("ParTcl"), Ruby ("Cardinal")
 + Update PIR, regex, Zcode
 + New language: Pynie - a Python compiler for Parrot
 + Lua implements require and many other ops, improved regex support
 + Remove parakeet
- Design:
 + PDD01 "Overview" - updated architecture and platform information
 + PDD15 "Objects" - details on roles, objects, and classes added
 + PDD22 "I/O" - added async ops and Status PMC details
- Documentation:
 + Added guides for Metacommitter, Relase Manager, and Cage Cleaner roles
- Implementation:
 + Object, Class, Role, and Attribute PMC implementation has begun
 + Perl 5 module "Parrot::Embed" now compiles and links on all platforms
- Build:
 + Major improvements in test coverage for 'ops2c.pl'
- Misc:
 + New utility: Crow, a template processor
 + New library module: Config/JSON for reading/writing JSON files
 + many bugfixes, enhancements, and coding standard updates, code cleanup

New in 0.4.9

- Compilers:
   + IMCC: Parrot calling conventions now available in C PMCs, allowing
     named, optional, slurpy, and flat parameter passing
   + PGE: extended support for Perl 5 Regexes
   + smop: prototype object model implementation
   + hllcompiler: refactored to run a configurable set of compilation stages
- PAST:
   + redesigned assign/binding to support Perl 6 binding semantics
- Languages:
   + Updated Lua, PHP ("Plumhead"), Tcl ("ParTcl"), perl6, perl5
   + New language: PIR - a PGE-based implementation of Parrot PIR
   + perl6 now supports binding (':=') and 'join'
   + lua generates tail calls, and supports its own regex flavor (PGE-based)
   + Pheme still works, huzzah!
- Design:
   + PDD21 "Objects" - rewritten
   + PDD22 "I/O" - updated and 'TODO' tests added
- Documentation:
   + Interface stability classification standards approved
   + Roles and Responsibilities documented approved
   + Official 'drafts' directory created (was 'clip')
- Implementation:
   + More NameSpace and OS PMC methods implemented
   + Parrot executable fullname and basename now available in PIR/PASM code
   + new 'chomp' library function
- Build:
   + Major improvements in test coverage for 'ops2pm.pl'
- Misc:
   + many bugfixes, enhancements, and coding standard updates
   + extended support for Sun Workshop Compilers
   + Parrot now builds on PocketPC platform

New in 0.4.8

- Compilers:
   + HLLCompiler: added tracing options, modified api
   + PGE & TGE bugfixes and updates
- PAST:
   + added global and lexical variable support
   + added looping constructs, arrays, hashes
- Languages:
   + Updated PHP ("Plumhead"), Tcl ("ParTcl"),
     forth, perl6, lua, abc, APL, WMLScript, punie
   + ParTcl is passing > 24.9% of Tcl cvs-latest test suite
   + perl6 now supports hashes, arrays, method calls, arity-based
     multisubs, quoted terms, ranges (non-lazy), try blocks, $!
- Design:

src/parrot_debugger.c
/*
Copyright (C) 2001-2008, Parrot Foundation.
$Id: parrot_debugger.c 37201 2009-03-08 12:07:48Z fperrad $

History:
    Renamed from pdb.c in 2008.7.15

=head1 NAME

parrot_debugger - The Parrot debugger

=head1 SYNOPSIS

 parrot_debugger programfile

=head1 DESCRIPTION

=head2 Commands

=over 4

=item C<disassemble>

Disassemble the bytecode.

Use this if you have a PBC file but not the PASM.

=item C<load>

Load a source code file.

=item C<list> or C<l>

List the source code file.

=item C<run> or C<r>

Run the program.

=item C<break> or C<b>

Add a breakpoint.

=item C<watch> or C<w>

Add a watchpoint.

=item C<delete> or C<d>

Delete a breakpoint.

=item C<disable>

Disable a breakpoint.

=item C<enable>

Reenable a disabled breakpoint.

=item C<continue> or C<c>

Continue the program execution.

=item C<next> or C<n>

Run the next instruction

=item C<eval> or C<e>

Run an instruction.

=item C<trace> or C<t>

Trace the next instruction.

=item C<print> or C<p>

Print the interpreter registers.

=item C<stack> or C<s>

Examine the stack.

=item C<info>

Print interpreter information.

=item C<quit> or C<q>

Exit the debugger.

=item C<help> or C<h>

Print the help.

=back

=head2 Debug Ops

You can also debug Parrot code by using the C<debug_init>, C<debug_load>
and C<debug_break> ops in F<ops/debug.ops>.

=over 4

=cut

*/

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include "../compilers/imcc/imc.h"
#include "../compilers/imcc/parser.h"
#include "parrot/embed.h"
#include "parrot/debugger.h"

static void PDB_printwelcome(void);
static void PDB_run_code(PARROT_INTERP, int argc, char *argv[]);

/*

=item C<int main(int argc, char *argv[])>

Reads the PASM or PBC file from argv[1], loads it, and then calls
Parrot_debug().

=cut

*/

int
main(int argc, char *argv[])
{
    int nextarg;
    Parrot_Interp     interp;

    /*
    Parrot_Interp     debugger = Parrot_new(interp);
    PDB_t            *pdb      = mem_allocate_zeroed_typed(PDB_t);
    */
    PDB_t *pdb;
    const char       *scriptname = NULL;
    const char       *filename;
    char             *ext;
    void             *yyscanner;

    Parrot_set_config_hash();

    interp = Parrot_new(NULL);

    Parrot_set_executable_name(interp, Parrot_str_new(interp, argv[0], 0));

    Parrot_debugger_init(interp);

    pdb = interp->pdb;

    /*Parrot_set_config_hash();  TODO link with cfg */

    pdb->state       = PDB_ENTER;

    Parrot_block_GC_mark(interp);
    Parrot_block_GC_sweep(interp);

    do_yylex_init(interp, &yyscanner);

    if (argc < 2) {
        fprintf(stderr, "Usage: parrot_debugger programfile [program-options]\n");
        Parrot_exit(interp, 1);
    }

    nextarg = 1;
    if (strcmp(argv[nextarg], "--script") == 0)
    {
        scriptname = argv [++nextarg];
        ++nextarg;
    }

    filename = argv[nextarg];
    ext      = strrchr(filename, '.');

    if (ext && STREQ(ext, ".pbc")) {
        Parrot_PackFile pf = Parrot_pbc_read(interp, filename, 0);

        if (!pf)
            return 1;

        Parrot_pbc_load(interp, pf);
        PackFile_fixup_subs(interp, PBC_MAIN, NULL);
    }
    else {
        Parrot_PackFile pf        = PackFile_new(interp, 0);
        int             pasm_file = 0;

        Parrot_pbc_load(interp, pf);

        IMCC_push_parser_state(interp);
        IMCC_INFO(interp)->state->file = filename;

        if (!(imc_yyin_set(fopen(filename, "r"), yyscanner)))    {
            IMCC_fatal_standalone(interp, EXCEPTION_PIO_ERROR,
                    "Error reading source file %s.\n",
                    filename);
        }

        if (ext && STREQ(ext, ".pasm"))
            pasm_file = 1;

        emit_open(interp, 1, NULL);
        IMCC_INFO(interp)->state->pasm_file = pasm_file;
        yyparse(yyscanner, interp);
        imc_compile_all_units(interp);

        imc_cleanup(interp, yyscanner);

        fclose(imc_yyin_get(yyscanner));
        PackFile_fixup_subs(interp, PBC_POSTCOMP, NULL);

        /* load the source for debugger list */
        PDB_load_source(interp, filename);

        PackFile_fixup_subs(interp, PBC_MAIN, NULL);
    }

    Parrot_unblock_GC_mark(interp);
    Parrot_unblock_GC_sweep(interp);

    if (scriptname)
        PDB_script_file(interp, scriptname);
    else
        PDB_printwelcome();

    interp->run_core = PARROT_DEBUGGER_CORE;
    PDB_run_code(interp, argc - nextarg, argv + nextarg);


    Parrot_exit(interp, 0);
}

/*

=item C<static void PDB_add_exception_handler(Parrot_Interp)>

Adds a default exception handler to PDB.

=cut

*/

static void
PDB_run_code(PARROT_INTERP, int argc, char *argv[])
{
    new_runloop_jump_point(interp);
    if (setjmp(interp->current_runloop->resume)) {
        free_runloop_jump_point(interp);
        fprintf(stderr, "Caught exception\n");
        return;
    }

    /* Loop to avoid exiting at program end */
    do {
        Parrot_runcode(interp, argc, argv);
        interp->pdb->state |= PDB_STOPPED;
    } while (! (interp->pdb->state & PDB_EXIT));
    free_runloop_jump_point(interp);
}

/*

=item C<static void PDB_printwelcome(void)>

Prints out the welcome string.

=cut

*/

static void
PDB_printwelcome(void)
{
    fprintf(stderr,
        "Parrot " PARROT_VERSION " Debugger\n"
        "\nPlease note: the debugger is currently under reconstruction\n");
}

/*

=back

=head1 SEE ALSO

F<src/debug.c>, F<include/parrot/debug.h>.

=head1 HISTORY

=over 4

=item * Initial version by Daniel Grunblatt on 2002.5.19.

=item * Start of rewrite - leo 2005.02.16

The debugger now uses it's own interpreter. User code is run in
Interp* debugee. We have:

  debug_interp->pdb->debugee->debugger
    ^                            |
    |                            v
    +------------- := -----------+

Debug commands are mostly run inside the C<debugger>. User code
runs of course in the C<debugee>.

=back

=head1 TODO

=over 4

=item * Check the user input for bad commands, it's quite easy to make
it bang now, try listing the source before loading or disassembling it.

=item * Print the interpreter info.

=item * Make the user interface better (add comands
history/completion).

=item * Some other things I don't remember now because it's late.


=back

=cut

*/


/*
 * Local variables:
 *   c-file-style: "parrot"
 * End:
 * vim: expandtab shiftwidth=4:
 */
